File: HOT100/7.链表/24.两两交换链表中的节点.cpp
/*
 * @lc app=leetcode.cn id=24 lang=cpp
 *
 * [24] 两两交换链表中的节点
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* pre = dummy;
        ListNode* cur = pre->next;
        if (! cur) return cur;
        ListNode* post = cur->next;
        while (post)
        {
            ListNode* tmp = post->next;
            pre->next = post;
            post->next = cur;
            cur->next = tmp;
            pre = cur;
            cur = pre->next;
            if (! cur) break;
            post = cur->next;
        }
        return dummy->next;
    }
};
// @lc code=end

File: HOT100/HOT100p4/24.两两交换链表中的节点.py
#
# @lc app=leetcode.cn id=24 lang=python3
#
# [24] 两两交换链表中的节点
#
from mytools import *
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        def rev(k):
            def rev_(node, k):
                pre = None
                cur = node.next
                for _ in range(k):
                    tmp = cur.next
                    cur.next = pre
                    pre = cur
                    cur = tmp
                ret = node.next
                node.next.next = cur
                node.next = pre
                return ret
            dummy = ListNode(next=head)
            n = 0
            p = head
            while p: n += 1; p = p.next
            p = dummy
            t = n // k
            for _ in range(t):
                p = rev_(p, k)
            return dummy.next
        return rev(2)
        
# @lc code=end

